library/VTKExtensions/Rendering/vtkF3DPolyDataMapper.cxx

```diff
@@ -130,25 +132,64 @@ void vtkF3DPolyDataMapper::ReplaceShaderValues(
   type = uniforms->GetUniformTupleType("jointMatrices");
   if (type != vtkUniforms::TupleTypeInvalid)
   {
-    customDecl += "in vec4 joints;\n"
-                  "in vec4 weights;\n";
-
-    // compute skinning matrix with current uniform weights
-    beginImpl += "  mat4 skinMat = weights.x * jointMatrices[int(joints.x)]\n"
-                 "               + weights.y * jointMatrices[int(joints.y)]\n"
-                 "               + weights.z * jointMatrices[int(joints.z)]\n"
-                 "               + weights.w * jointMatrices[int(joints.w)];\n";
+    int nbJoints = uniforms->GetUniformNumberOfTuples("jointMatrices");
 
-    posImpl += "  posMC = skinMat * posMC;\n";
+    bool skinningSupported = true;
 
-    // apply the matrix to normals and tangents
-    if (hasNormals)
+    // The number of uniform values required by OpenGL is 4096
+    // Since a mat4 is 16 values, it means we can only support 256 bones
+    // However, there are other uniform values used in practice for other
+    // things like materials and we've seen issues starting at 253 bones
+    // let's be conservative here and trigger SSBO at 250 bones
+    if (nbJoints > 250)
     {
-      normalImpl += "  normalVCVSOutput = mat3(skinMat) * normalVCVSOutput;\n";
+#if VTK_VERSION_NUMBER >= VTK_VERSION_CHECK(9, 3, 20231108)
+      std::vector<float> buffer(16 * sizeof(float) * nbJoints);
+      uniforms->GetUniformMatrix4x4v("jointMatrices", buffer);
+
+      this->JointMatrices->Upload(buffer, vtkOpenGLBufferObject::ArrayBuffer);
+      this->JointMatrices->BindShaderStorage(0);
+
+      uniforms->RemoveUniform("jointMatrices");
+
+      customDecl += "layout(std430, binding = 0) readonly buffer JointsData\n"
+                    "{\n"
+                    "  mat4 jointMatrices[];\n"
+                    "};\n";
+
+      // 4.3 is required for SSBO
+      vtkShaderProgram::Substitute(VSSource, "//VTK::System::Dec", "#version 430");
+#else
+      std::string msg = "A mesh is associated with more than 250 bones (" +
+        std::to_string(nbJoints) + "), which is not supported by VTK <= 9.3";
```

Comment:
9.3 is not precise enough, it should say 9.4 or 9.3.20231108 I think

Reply:
I'll put `9.3.20231108` as we're not sure if VTK next release will be 9.4 or 10.0 I guess.